core/blockchain/blockchain_gateway.py

```diff
@@ -0,0 +1,355 @@
+        and put into the optimizer initially. So the optimizer knows this info.
+        """
+        logging.info("handling decentralized learning... {}".format(tx))
+        key = tx.get('key')
+        value = tx.get('content')
+        args = {'key': key, 'content': self._ipfs_to_content(value)}
+        self.communication_manager.inform("new_session", args)
+        self.listen_new_weights()
+
+    def handle_new_weights(self, tx: dict):
+        """
+        handle_new_weights() method downloads weights and does something with
+        it. This callback is triggered by the Listener when it sees new weights 
+        intended for its node ID. Downloads the weights, and gives them to the
+        comm. mgr which gives them to the relevant optimizer 
+        -which should do the moving average.
+        """
+        logging.info("handling new weights...{}".format(tx))
+        key = tx.get('key')
+        value = tx.get('content')
+        args = {'key': key, 'content': self._ipfs_to_content(value)}
+        # TODO: Put into in-memory datastore.   
+        self.communication_manager.inform("new_weights", args)
+        self.listen_new_weights()
+
+    def handle_terminate(self):
+        self.communication_manager.inform("TERMINATE", None)
+
+    def listen_decentralized_learning(self):
+        """
+        Polls blockchain for node ID in decentralized_learning() method
+        signature decentralized_learning(...<node_ids>...) method signature
+        will be the key on the blockchain; listener should look for this, and
+        if the method signature contains its node id, it will trigger a
+        callback
+        """
+        def filter(tx):
+            logging.info("tx: {}".format(tx))
+            return tx.get('key') == tx.get('content')
+        return self.filter_set(filter,
+                                self.handle_decentralized_learning_trainer)
+
+    def broadcast_new_weights(self, tx: dict) -> str:
+        """
+        broadcast_new_weights() method with all relevant parameters
+        should populate the key of new_weights with all of the nodes for which
+        these new weights are relevant. value should be IPFS hash.
+        """
+        content = tx.get('content', None)
+        key = self.client.add_json(content['weights'])
+        tx_receipt = self.setter(key, content['gradient'])
+        return tx_receipt
+
+    def listen_new_weights(self):
+        """
+        Polls blockchain for node ID in new_weights() method signature
+        new_weights(...<node_ids>...) method signature will be the key on the
+        blockchain; listener should look for this, and if the method signature
+        contains its node id, it will trigger a callback
+        """
+        logging.info("I'm listening for new weights!")
+        def weights_filter(tx):
+            logging.info("filtering for new weights: {}".format(tx))
+            return tx.get('key') != tx.get('content')
```

Comment:
Same as above.

Reply:
same as above? what is same as above

Reply:
The .get() failing. Assert works but be aware of where this will be used -- the assert could have been done already, in which case you just need to comment that in the docstring.